runtime/eventring.c

```diff
@@ -0,0 +1,549 @@
+      }
+
+      current_metadata = MapViewOfFile(
+        ring_handle,
+        FILE_MAP_ALL_ACCESS,
+        0,
+        0,
+        0
+      );
+
+      if( current_metadata == NULL ) {
+        caml_fatal_error("failed to map view of file");
+      }
+#else
+      ring_fd =
+          open(current_ring_loc, O_RDWR | O_CREAT, (S_IRUSR | S_IWUSR));
+
+      if (ring_fd < 0) {
+        caml_fatal_error("Couldn't open ring buffer loc: %s",
+                         current_ring_loc);
+      }
+
+      ret = ftruncate(ring_fd, current_ring_total_size);
+
+      if (ret < 0) {
+        caml_fatal_error("Can't resize ring buffer");
+      }
+
+      current_metadata = mmap(NULL, current_ring_total_size,
+                              PROT_READ | PROT_WRITE, MAP_SHARED, ring_fd, 0);
+
+      if (current_metadata == NULL) {
+        caml_fatal_error("Unable to mmap ring buffer");
+      }
+
+      close(ring_fd);
+#endif
+      ring_headers_length =
+          Max_domains * sizeof(struct eventring_buffer_header);
+
+      current_metadata->version = EVENTRING_VERSION;
+      current_metadata->max_domains = Max_domains;
+      current_metadata->ring_header_size_bytes =
+          sizeof(struct eventring_buffer_header);
+      current_metadata->ring_size_bytes =
+          ring_size_words * sizeof(uint64_t);
+      current_metadata->ring_size_elements = ring_size_words;
+      current_metadata->headers_offset =
+          sizeof(struct eventring_metadata_header);
+
+      /* strictly we can calculate this in a consumer but for simplicity we
+         store it in the metadata header */
+      current_metadata->data_offset =
+        current_metadata->headers_offset + ring_headers_length;
+
+      for (int domain_num = 0; domain_num < Max_domains; domain_num++) {
+        /* we initialise each ring's metadata. We use the offset to the headers
+          and then find the slot for each of domain in Max_domains */
+        struct eventring_buffer_header *ring_buffer =
+            (struct eventring_buffer_header
+                 *)((char *)current_metadata +
+                    current_metadata->headers_offset +
+                    domain_num * sizeof(struct eventring_buffer_header));
+
+        ring_buffer->ring_head = 0;
+        ring_buffer->ring_tail = 0;
+      }
+
+      atomic_store_rel(&eventring_enabled, 1);
+      atomic_store_rel(&eventring_paused, 0);
+
+      caml_ev_lifecycle(EV_RING_START, pid);
+    }
+  }
+  caml_global_barrier();
+}
+
+CAMLprim value caml_eventring_start() {
+  while (!atomic_load_acq(&eventring_enabled)) {
+    caml_try_run_on_all_domains(&stw_create_eventring, NULL, NULL);
+  }
+
+  return Val_unit;
+}
+
+CAMLprim value caml_eventring_pause() {
+  uintnat paused = atomic_load_acq(&eventring_paused);
+
+  if (atomic_load_acq(&eventring_enabled) && !paused) {
+    if( atomic_compare_exchange_strong(&eventring_paused, &paused, 1) ) {
+      caml_ev_lifecycle(EV_RING_PAUSE, 0);
+    }
+  }
```

Comment:
I think that there is no need to pre-read the `&eventring_paused` value before the compare-and-set, so this could be simplified into something like:

```c
  if (!atomic_load_acq(&eventring_enabled)) return Val_unit;
  uintnat not_paused = 0;
  if( atomic_compare_exchange_strong(&eventring_paused, &not_paused, 1) ) {
    caml_ev_lifecycle(EV_RING_PAUSE, 0);
  }
  return Val_unit;
```

(Same with `resume` below.)

Reply:
Done in https://github.com/ocaml/ocaml/pull/10964/commits/f25b5765d2f5c612428b6f1edb5a35eeb48148ca